### ISSUE TITLE
"Verification failed: Struct field type mismatch" compiler error

### ISSUE THREAD

**Issue by dmihal** [FuelLabs/sway#5492]
The following code is generating the error "Internal compiler error: Verification failed: Struct field type mismatch."

```rust
contract;

use std::{hash::Hash};

storage {
    total_supply: StorageMap<AssetId, u64> = StorageMap {},
}

abi SRC20 {
    #[storage(read)]
    fn decimals(asset: AssetId) -> Option<u8>;
}

impl SRC20 for Contract {
    #[storage(read)]
    fn decimals(asset: AssetId) -> Option<u8> {
        match storage.total_supply.get(asset).try_read() {
            Option::Some(_supply) => Option::Some(0),
            Option::None => Option::None,
        }
    }
}
```

Compiled with `forc 0.48.1+nightly.20231226.a676a3d981`

**Comment by vaivaswatha** [comment]
Simplified examples

```
script;

fn main() {
  let _ = bar(true);
}

#[inline(never)]
fn bar(b: bool) -> Option<u8> {
   if(b) {
     Option::Some(19)
   } else {
     Option::None
   }
}
```

```
script;

fn main() {
  let _ = foo();
}

#[inline(never)]
fn foo() -> Option<u8> {
  match Some(true) {
    Option::Some(_b) => Option::Some(17),
    Option::None => Option::None,
  }
}
```

**Comment by ironcev** [comment]
Blocked by #5562 (only the first two tasks are needed to properly solve this issue).

If urgent, we can provide a workaround and fix this issue before getting the first two tasks of #5562 done.

The workaround would suppress the type unification in case of a `Never` type when type checking `match` branches.

This should solve the issue, but it is an ugly hack in the compiler code. If the issue is not urgent, I am for waiting until the first two tasks of #5562 are done.

**Comment by ironcev** [comment]
Closed in #5643.
